lib/bandit/http2/stream.ex

```diff
@@ -100,140 +80,23 @@ defmodule Bandit.HTTP2.Stream do
-    end)
-    |> if do
-      :ok
-    else
-      {:error, {:stream, stream_id, Errors.protocol_error(), "Received invalid pseudo header"}}
-    end
-  end
-
   # RFC7540§8.1.2.1 - pseudo headers must appear first
   defp no_pseudo_headers(headers, stream_id) do
-    headers
-    |> Enum.any?(fn {key, _value} -> String.starts_with?(key, ":") end)
-    |> if do
+    if Enum.any?(headers, fn {key, _value} -> String.starts_with?(key, ":") end) do
       {:error,
        {:stream, stream_id, Errors.protocol_error(), "Received trailers with pseudo headers"}}
     else
       :ok
     end
   end
 
-  # RFC7540§8.1.2.2 - pseudo headers must appear first
-  defp pseudo_headers_first(headers, stream_id) do
-    headers
-    |> Enum.drop_while(fn {key, _value} -> String.starts_with?(key, ":") end)
-    |> Enum.any?(fn {key, _value} -> String.starts_with?(key, ":") end)
-    |> if do
-      {:error,
-       {:stream, stream_id, Errors.protocol_error(), "Received pseudo headers after regular one"}}
-    else
-      :ok
-    end
-  end
-
-  # RFC7540§8.1.2.2 - no hop-by-hop headers from RFC2616§13.5.1
-  # Note that we do not filter out the TE header here, since it is allowed in
-  # specific cases by RFC7540§8.1.2.2. We check those cases in a separate filter
-  defp no_connection_headers(headers, stream_id) do
-    headers
-    |> Enum.any?(fn {key, _value} ->
-      key in ~w[connection keep-alive proxy-authenticate proxy-authorization trailers transfer-encoding upgrade]
-    end)
-    |> if do
-      {:error,
-       {:stream, stream_id, Errors.protocol_error(), "Received connection-specific header"}}
-    else
-      :ok
-    end
-  end
-
-  # RFC7540§8.1.2.2 - TE header may be present if it contains exactly 'trailers'
-  defp valid_te_header(headers, stream_id) do
-    case List.keyfind(headers, "te", 0) do
-      nil ->
-        :ok
-
-      {_, "trailers"} ->
-        :ok
-
-      _ ->
-        {:error, {:stream, stream_id, Errors.protocol_error(), "Received invalid TE header"}}
-    end
-  end
-
-  # RFC7540§8.1.2.3 - method, scheme, path pseudo headers must appear exactly once
-  defp exactly_one_instance_of(headers, header, stream_id) do
-    headers
-    |> Enum.count(fn {key, _value} -> key == header end)
-    |> case do
-      1 ->
-        :ok
-
-      _ ->
-        {:error, {:stream, stream_id, Errors.protocol_error(), "Expected 1 #{header} headers"}}
-    end
-  end
-
-  # RFC7540§8.1.2.3 :path must not be empty
-  defp non_empty_path(headers, stream_id) do
-    case List.keyfind(headers, ":path", 0) do
-      {_, ""} ->
-        {:error, {:stream, stream_id, Errors.protocol_error(), "Received empty :path"}}
-
-      _ ->
-        :ok
-    end
-  end
-
   defp expected_content_length(headers) do
     case List.keyfind(headers, "content-length", 0) do
-      nil -> nil
-      {_, content_length} -> String.to_integer(content_length)
+      nil -> {:ok, nil}
+      {_, content_length} -> {:ok, String.to_integer(content_length)}
```

Comment:
Should you do another `case` on `Integer.parse`, or is raising here expected if it doesn't parse?

Reply:
This is a WiP - I'm going to DRY up @moogle19's recent work on content-length parsing from HTTP/1 and call it here. I wan to make sure that the containing plumbing supports the `{:ok, ...}` wrapping is all.